datastore/datastore.go

```diff
@@ -0,0 +1,217 @@
+)
+
+//DataStore exported
+type DataStore interface {
+	// PutObject adds a new Record based on an object into the datastore
+	PutObject(kvObject KV) error
+	// PutObjectAtomic provides an atomic add and update operation for a Record
+	PutObjectAtomic(kvObject KV) error
+	// KVStore returns access to the KV Store
+	KVStore() store.Store
+}
+
+type datastore struct {
+	store store.Store
+}
+
+//KV Key Value interface used by objects to be part of the DataStore
+type KV interface {
+	Key() []string
+	Value() []byte
+	Index() uint64
+	SetIndex(uint64)
+}
+
+//Key provides convenient method to create a Key
+func Key(key ...string) string {
+	keychain := []string{"docker", "libnetwork"}
+	keychain = append(keychain, key...)
+	str := strings.Join(keychain, "/")
+	return str + "/"
+}
+
+var errNewDatastore = errors.New("Error creating new Datastore")
+var errInvalidConfiguration = errors.New("Invalid Configuration passed to Datastore")
+
+// newClient used to connect to KV Store
+func newClient(kv string, addrs string) (DataStore, error) {
+	store, err := store.CreateStore(kv, []string{addrs}, store.Config{})
+	if err != nil {
+		return nil, err
+	}
+	ds := &datastore{store: store}
+	return ds, nil
+}
+
+// NewDataStore creates a new instance of LibKV data store
+func NewDataStore(cfg *config.DatastoreCfg) (DataStore, error) {
+	if cfg == nil {
+		return nil, errInvalidConfiguration
+	}
+	// TODO : cfg.Embedded case
+	return newClient(cfg.Client.Provider, cfg.Client.Address)
+}
+
+// NewCustomDataStore can be used by clients to plugin cusom datatore that adhers to store.Store
+func NewCustomDataStore(customStore store.Store) DataStore {
+	return &datastore{store: customStore}
+}
+
+func (ds *datastore) KVStore() store.Store {
+	return ds.store
+}
+
+// PutObjectAtomic adds a new Record based on an object into the datastore
+func (ds *datastore) PutObjectAtomic(kvObject KV) error {
+	if kvObject == nil {
+		return errors.New("kvObject is nil")
+	}
+	kvObjValue := kvObject.Value()
+
+	if kvObjValue == nil {
+		return ErrInvalidAtomicRequest
+	}
+	_, err := ds.store.AtomicPut(Key(kvObject.Key()...), []byte{}, kvObjValue, kvObject.Index())
+	if err != nil {
+		return err
+	}
+
+	_, index, err := ds.store.Get(Key(kvObject.Key()...))
+	if err != nil {
+		return err
+	}
+	kvObject.SetIndex(index)
+	return nil
+}
+
+// PutObject adds a new Record based on an object into the datastore
+func (ds *datastore) PutObject(kvObject KV) error {
+	if kvObject == nil {
+		return errors.New("kvObject is nil")
+	}
+	return ds.putObjectWithKey(kvObject, kvObject.Key()...)
+}
+
+func (ds *datastore) putObjectWithKey(kvObject KV, key ...string) error {
+	kvObjValue := kvObject.Value()
+
+	// If the KVObject provides a Value, use it as is. No further processing required.
+	if kvObjValue != nil {
+		return ds.store.Put(Key(key...), kvObjValue)
+	}
+
+	// Else, DataStore will rely on `kv:"<tag>"` Tag to form the KV pair(s) for the object
+	var keychain []string
```

Comment:
Are we using this facility? Can we remove the support for it if we are not using it? It reduces the review complexity at this time given we have a lot to review :-)